charms/knative-serving/src/lightkube_helpers/__init__.py

```diff
@@ -0,0 +1,88 @@
+from typing import List, Union, Tuple
+
+import lightkube
+from lightkube.core.resource import NamespacedResource, GlobalResource
+from lightkube.resources.apps_v1 import StatefulSet
+from ops.model import ActiveStatus, WaitingStatus, BlockedStatus, MaintenanceStatus
+
+from charm_helpers.exceptions import ErrorWithStatus
+
+
+def check_resources(client: lightkube.Client, expected_resources: List[Union[NamespacedResource, GlobalResource]]) -> \
+        (bool, List[Exception]):
+    """Checks status of resources in cluster, returning True if all are considered ready
+
+    Note: This is a basic skeleton of a true check on the resources.  Currently it only checks
+    that resources exist and that StatefulSets have their desired number of replicas ready.
+
+    Returns: Tuple of:
+        Status (bool)
+        List of Exceptions encountered during failed checks, with each entry
+        indexed the same as the corresponding expected_resource (list[str])
+    """
+    errors = [None] * len(expected_resources)
+    for i, expected_resource in enumerate(expected_resources):
+        try:
+            found_resource = client.get(
+                type(expected_resource),
+                expected_resource.metadata.name,
+                namespace=expected_resource.metadata.namespace,
+            )
+        except lightkube.core.exceptions.ApiError as e:
+            msg = f"Cannot find k8s object corresponding to '{expected_resource.metadata}'"
+            errors[i] = ResourceNotFoundError(msg, BlockedStatus)
+            continue
+
+        if isinstance(found_resource, StatefulSet):
+            try:
+                validate_statefulset(found_resource)
+            except ReplicasNotReadyError as e:
+                errors[i] = e
+
+    return not any(errors), errors
+
+
+def validate_statefulset(resource: StatefulSet) -> bool:
```

Comment:
Why are we specifically checking for statefulsets here?

Reply:
I think it would make sense to check for more things, but I don't have code for those.  

Things that have no workload and should just exist (CRDs, CMs) are easy to check (they're caught by the `.get` check), but other objects need more inspection.  I had this code to inspect StatefulSets from a previous charm so I've included it here.  I think the logic would apply to some other things too that we could add